### PR TITLE
evalengine: Refactorings & fixes

### PR DESCRIPTION
**Pull request by vmg**
## Description

While working on https://github.com/vitessio/vitess/pull/12369, @dbussink and I have found several bugs affecting the evaluator (not the compiler), and several cases of duplicated code that needed to be extracted in order to support having a compiler and evaluator. 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by dbussink — APPROVED** [review]

File: go/vt/vtgate/evalengine/api_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"testing"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+func parseExpr(t testing.TB, expr string) sqlparser.Expr {
+	stmt, err := sqlparser.Parse("select " + expr)
```
Comment (systay) [inline go/vt/vtgate/evalengine/api_test.go:13]: we already have `sqlparser.ParseExpr` that does this

File: go/vt/vtgate/evalengine/fn_bit.go [hunk 2]
```diff
@@ -52,5 +52,6 @@ func (call *builtinBitCount) eval(env *ExpressionEnv) (eval, error) {
 
 func (call *builtinBitCount) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
 	_, f := call.Arguments[0].typeof(env)
+	// The MySQL docs are actually wrong and this returns an int64, not a uint64.
```
Comment (systay) [inline go/vt/vtgate/evalengine/fn_bit.go:55]: 😭 😆

File: go/vt/vtgate/evalengine/testcases/helpers.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package testcases
```
Comment (systay) [inline go/vt/vtgate/evalengine/testcases/helpers.go:1]: license header missing

File: go/vt/vtgate/evalengine/testcases/inputs.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package testcases
```
Comment (systay) [inline go/vt/vtgate/evalengine/testcases/inputs.go:1]: license

**Review by systay — APPROVED** [review]

**Pull request merged**